### DISCUSSION TITLE
Is it possible to set alarms in XTCE MDB on a per-parameter basis ?

### DISCUSSION THREAD

**Discussion by sasataku** [yamcs/yamcs#893] [Q&A] [answered]
In the MDB Excel format, I believe we can set alarms for each `parameter`.
However, in the case of XTCE, it seems that alarms can only be set for each `parameter type`.

For example, I define a `parameter type` to represent `Shunt voltage` with unit and calibration.

```
<FloatParameterType sizeInBits="32" name="shunt">
        <UnitSet>
                <Unit>mA</Unit>
        </UnitSet>
        <IntegerDataEncoding encoding="unsigned" byteOrder="leastSignificantByteFirst" sizeInBits="32">
                <DefaultCalibrator>
                        <PolynomialCalibrator>
                                <Term exponent="1" coefficient="0.1" />
                        </PolynomialCalibrator>
                </DefaultCalibrator>
        </IntegerDataEncoding>
</FloatParameterType>
```
And this parameter type is referenced by multiple parameters.

```
<Parameter parameterTypeRef="/basetype/shunt" name="1V8_Shunt" />
<Parameter parameterTypeRef="/basetype/shunt" name="3V3_Shunt" />
```
I want to set alarms for each parameter, but since alarms cannot be set for parameters themselves, I need to add alarms to the `parameter type` and create multiple parameter types.

```
<FloatParameterType sizeInBits="32" name="shunt_1v8">
       <DefaultAlarm>
                <StaticAlarmRanges>
                      <WarningRange minInclusive="0.0" maxInclusive="100.0" />
                 </StaticAlarmRanges>
       </DefaultAlarm>
        <UnitSet>
                <Unit>mA</Unit>
        </UnitSet>
        <IntegerDataEncoding encoding="unsigned" byteOrder="leastSignificantByteFirst" sizeInBits="32">
                <DefaultCalibrator>
                        <PolynomialCalibrator>
                                <Term exponent="1" coefficient="0.1" />
                        </PolynomialCalibrator>
                </DefaultCalibrator>
        </IntegerDataEncoding>
</FloatParameterType>

<FloatParameterType sizeInBits="32" name="shunt_3v3">
       <DefaultAlarm>
                <StaticAlarmRanges>
                      <WarningRange minInclusive="0.0" maxInclusive="300.0" />
                 </StaticAlarmRanges>
       </DefaultAlarm>
        <UnitSet>
                <Unit>mA</Unit>
        </UnitSet>
        <IntegerDataEncoding encoding="unsigned" byteOrder="leastSignificantByteFirst" sizeInBits="32">
                <DefaultCalibrator>
                        <PolynomialCalibrator>
                                <Term exponent="1" coefficient="0.1" />
                        </PolynomialCalibrator>
                </DefaultCalibrator>
        </IntegerDataEncoding>
</FloatParameterType>
```

In our system, we want to differentiate alarms based on the measurement location, even if the calibration and units are the same.
Rather than creating multiple parameter types, is it possible to overwrite the default alarm of the parameter type for each parameter?

**Comment by xpromache** [comment, marked as answer]
Unfortunately not possible, that's how XTCE defines it.

But when defining a type, you can use the attribute "baseType" to inherit
the properties of another type.
So you can have one type with the calibrations and encoding and then one
additional type for each alarm configuration.

Something like this:

<FloatParameterType sizeInBits="32" name="shunt_3v3">
        <UnitSet>
                <Unit>mA</Unit>
        </UnitSet>
        <IntegerDataEncoding encoding="unsigned"
byteOrder="leastSignificantByteFirst" sizeInBits="32">
                <DefaultCalibrator>
                        <PolynomialCalibrator>
                                <Term exponent="1" coefficient="0.1" />
                        </PolynomialCalibrator>
                </DefaultCalibrator>
        </IntegerDataEncoding>
</FloatParameterType>

<FloatParameterType name="shunt_3v3_with_alarm1" baseType="shunt_3v3">
       <DefaultAlarm>
                <StaticAlarmRanges>
                      <WarningRange minInclusive="0.0" maxInclusive="300.0" />
                 </StaticAlarmRanges>
       </DefaultAlarm>

</FloatParameterType>

<Parameter parameterTypeRef="/basetype/shunt_3v3_with_alarm1"
name="3V3_Shunt" />

**Comment by sasataku** [comment]
Thanks for the answer, Nicolae.
I didn't know about `baseType` !
This means we won't need to define duplicate unit/calibration definitions. 
It's helpful to me.

Thanks, I will close the discussion.
